### DISCUSSION TITLE
Use HMP4040 for other HMP models?

### DISCUSSION THREAD

**Discussion by annkvth** [pymeasure/pymeasure#1102] [Q&A]
Hi all,

I am considering to use pymeasure to control an HMP2020 power supply. I discovered that there is code for a related model, the 4040.  

As far as I can tell by looking at the R&S documentation, the only difference between the **HMP2020**, _HMP2030_, HMP4030 and **HMP4040** are the number of output channel (**2**/_3_/3/**4**) and the maximum output current per channel (**5A-and-10A-mixed**/_5A_/10A/**10A**). 

So I was wondering what the best way for me to work with the 2020 would be:

- just use the HMP4040 instrument class and keep in mind that I don't have a channel 3 and 4
- make a copy of the existing file `pymeasure/instruments/rohdeschwarz/hmp.py` and adapt it (e.g. for the channel selection commands, where selected_channel now contains `values=[1, 2, 3, 4]`, replace that with `values=[1, 2]`)
- somehow(???) edit the existing class to adapt to different HMP models - that would avoid duplicating a lot of code, but at first glance, it's not clear to me how I would go about doing that. 

I looked at the "adding instruments" guide in the documentation, but it didn't become clear to me what the preferred strategy for instrument families with small differences is.

**Comment by BenediktBurger** [comment]
Hey @annkvth ,

thank you very much for your question and readiness to share the code with the community.

If there are only tiny differences, the best option is, to create a base class for all the similar instruments and then you create tiny classes inheriting from that base class.
If I understand it correctly, you could basically take the HMP4040 as the base class.

Dynamic properties are really great for adjusting the limits in different classes, which you can find [here in the documentation](https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments/properties.html#instruments-with-similar-features).

For example, you would add `dynamic=True` to the select channels and in the init of your HMP2020 subclass, you set `selected_channels_values = [1, 2]`. Note that there is an underscore, no period, between `selected_channels` and `values`. That's the _magic_ of dynamic properties 😁 .

If you have questions, don't hesitate to ask!

**Reply by annkvth** [reply]
Thanks! I don't know how I managed to overlook the "Instruments with similar features" section in the documentation before :sweat_smile: 

I haven't used dynamic properties in python before, but that sounds very useful - I'll give it a try!

**Reply by BenediktBurger** [reply]
These dynamic properties are a pymeasure creation, not plain python properties.

The are quite useful, but not that obvious, how they work.
